_source/_posts/2024-02-27-spring-security-pkce.md

```diff
@@ -0,0 +1,458 @@
+
+> Since version 2.1.6, the Okta Starter enables PKCE by default for confidential clients. With the Okta Starter default auto-configuration, PKCE is enabled automatically even if the client-secret is set through the Okta Starter configuration properties.
+
+In the browser window, continue with the sign-in flow, and give consent to the application to access your user information:
+
+{% img blog/spring-security-pkce/auth0-consent.png alt:"Auth0 Consent Page" width:"400" %}{: .center-image }
+
+After the approval, Auth0 will redirect the browser to the `index.html` page. At the top right, the navigation has a user drop-down menu, with **Profile** and **Logout** options. If you click on **Profile**, it will display the ID Token claims. If you click on **Logout**, it will end the local session and redirect to a _logged out_ page generated by Spring Security.
+
+{% img blog/spring-security-pkce/home-page.png alt:"Application Home Page" width:"800" %}{: .center-image }
+
+### Configure the logout
+
+Modify the logout flow with the following configuration:
+
+```yml
+# src/main/resources/application.yml
+...
+okta:
+  oauth2:
+    ...
+    post-logout-redirect-uri: "{baseUrl}"
+...    
+```
+
+Spring security will resolve `{baseUrl}` placeholder to the application base URL at request time.
+
+Make sure [End Session Endpoint Discovery](https://auth0.com/docs/authenticate/login/logout/log-users-out-of-auth0#enable-endpoint-discovery) is enabled in your Auth0 tenant. If enabled, Spring Security will logout the user at the provider with [RP-Initated Logout](https://auth0.com/docs/authenticate/login/logout/log-users-out-of-auth0). In the Auth0 dashboard, choose the **Settings** option from the left menu, and then click the **Advanced** tab.
+
+{% img blog/spring-security-pkce/auth0-end-session-discovery.png alt:"Auth0 End Session Endpoint Discovery toggle option" width:"600" %}{: .center-image }
+
+Restart the application. Now the **Logout** link will end the session at Auth0, and the browser will redirect to the Universal Login page.
+
+### Enable PKCE for confidential clients with custom HttpSecurity
+
+You can customize the OIDC login by defining your own web security, adding a custom `LogoutSuccessHandler` for making the application redirect to the base URL after the logout.
+
+> If the default security configuration is customized with `HttpSecurity.oauth2Login()`, you can re-enable PKCE for confidential clients with [`OAuth2AuthorizationRequestCustomizers.withPkce()`](https://docs.spring.io/spring-security/reference/servlet/oauth2/client/authorization-grants.html#_obtaining_authorization) or with the utility method `Okta.configureOAuth2WithPkce()`
+
+The Okta Starter auto-configuration is conditional to the application not defining its own web security, so when customizing with `HttpSecurity.oauth2Login()`, PKCE is not enabled by default for confidential clients (clients with a secret), and you must enable it explicitly.
```

Comment:
can you also mention how to enable PKCE with spring boot if not using okta starter